src/main/scala/io/iohk/ethereum/consensus/pow/EthashMiner.scala

```diff
@@ -0,0 +1,205 @@
+    case ProcessMining => processMining()
+  }
+
+  def processMining(): Unit = {
+    blockchain.getBestBlock() match {
+      case Some(blockValue) =>
+        blockCreator
+          .getBlockForMining(blockValue)
+          .map { case PendingBlockAndState(PendingBlock(block, _), _) =>
+            val blockNumber = block.header.number
+            val (startTime, mineResult) =
+              // ECIP-1049 //TODO refactoring in ETCM-759 to remove the if clause
+              if (isKeccak(blockNumber)) doKeccakMining(block, blockCreator.miningConfig.mineRounds)
+              else doEthashMining(blockNumber.toLong, block)
+
+            val time = System.nanoTime() - startTime
+            //FIXME: consider not reporting hash rate when time delta is zero
+            val hashRate = if (time > 0) (mineResult.triedHashes.toLong * 1000000000) / time else Long.MaxValue
+            ethMiningService.submitHashRate(SubmitHashRateRequest(hashRate, ByteString("mantis-miner")))
+            mineResult match {
+              case MiningSuccessful(_, mixHash, nonce) =>
+                log.info(
+                  s"Mining successful with ${ByteStringUtils.hash2string(mixHash)} and nonce ${ByteStringUtils.hash2string(nonce)}"
+                )
+                syncController ! SyncProtocol.MinedBlock(
+                  block.copy(header = block.header.copy(nonce = nonce, mixHash = mixHash))
+                )
+              case _ => log.info("Mining unsuccessful")
+            }
+            self ! ProcessMining
+          }
+          .onErrorHandle { ex =>
+            log.error(ex, "Unable to get block for mining")
+            context.system.scheduler.scheduleOnce(10.seconds, self, ProcessMining)
+          }
+          .runAsyncAndForget
+      case None =>
+        log.error("Unable to get block for mining, getBestBlock() returned None")
+        context.system.scheduler.scheduleOnce(10.seconds, self, ProcessMining)
+    }
+  }
+
+  private def isKeccak(currentBlockNumber: BigInt): Boolean = {
+    ecip1049BlockNumber match {
+      case None => false
+      case Some(blockNumber) => currentBlockNumber >= blockNumber
+    }
+  }
+
+  private def doEthashMining(blockNumber: Long, block: Block): (Long, MiningResult) = {
+    val epoch =
+      EthashUtils.epoch(blockNumber, blockCreator.blockchainConfig.ecip1099BlockNumber.toLong)
+    val (dag, dagSize) = dagManager.calculateDagSize(blockNumber, epoch)
+    val headerHash = crypto.kec256(BlockHeader.getEncodedWithoutNonce(block.header))
+    val startTime = System.nanoTime()
+    val mineResult =
+      mineEthah(headerHash, block.header.difficulty.toLong, dagSize, dag, blockCreator.miningConfig.mineRounds)
+    (startTime, mineResult)
+  }
+
+  private def doKeccakMining(block: Block, numRounds: Int): (Long, MiningResult) = {
+    val rlpEncodedHeader = BlockHeader.getEncodedWithoutNonce(block.header)
+    val initNonce = BigInt(64, new Random()) // scalastyle:ignore magic.number
```

Comment:
maybe turn 64 into a 'not magic' number instead of omitting the warning?

Reply:
I am placing 64 in a val, in ETCM-759, that touches this class